packages/gatsby-codemods/src/transforms/gatsby-source-contentful.js

```diff
@@ -0,0 +1,291 @@
+            )
+            state.opts.hasChanged = true
+          }
+        }
+      },
+    },
+  }
+}
+
+const RENAME_BLACKLIST = [`Asset`, `Reference`]
+const SYS_FIELDS_TRANSFORMS = new Map([
+  [`node_locale`, `locale`],
+  [`contentful_id`, `id`],
+  [`spaceId`, `spaceId`],
+  [`createdAt`, `firstPublishedAt`],
+  [`updatedAt`, `publishedAt`],
+  [`revision`, `publishedVersion`],
+  [`type`, `contentType`], // @todo special case, is now a link
+])
+
+function locateSubfield(node, fieldName) {
+  return (
+    node.selectionSet &&
+    node.selectionSet.selections.find(({ name }) => name?.value === fieldName)
+  )
+}
+
+function processGraphQLQuery(query, state) {
+  try {
+    let hasChanged = false // this is sort of a hack, but print changes formatting and we only want to use it when we have to
+    const ast = graphql.parse(query)
+
+    graphql.visit(ast, {
+      SelectionSet(node) {
+        // Rename content type node selectors
+        const contentfulSelector = node.selections.find(({ name }) => {
+          const res = name?.value.match(
+            /^(allContentful|contentful)([A-Z0-9].+)/
+          )
+          return res && !RENAME_BLACKLIST.includes(res[2])
+        })
+
+        if (contentfulSelector) {
+          contentfulSelector.name.value = contentfulSelector.name.value.replace(
+            `ontentful`,
+            `ontentfulContentType`
+          )
+          hasChanged = true
+          return
+        }
+
+        // Move sys attributes into real sys
+        const contentfulSysFields = node.selections.filter(({ name }) =>
+          SYS_FIELDS_TRANSFORMS.has(name?.value)
+        )
+
+        if (contentfulSysFields.length) {
+          const transformedSysFields = cloneDeep(contentfulSysFields).map(
+            field => {
+              return {
+                ...field,
+                name: {
+                  ...field.name,
+                  value: SYS_FIELDS_TRANSFORMS.get(field.name.value),
+                },
+              }
+            }
+          )
+
+          // Replace first old field occurence with new sys field
+          let sysInjected = false
+          const sysField = {
+            kind: `Field`,
+            name: {
+              kind: `Name`,
+              value: `sys`,
+            },
+            selectionSet: {
+              kind: `SelectionSet`,
+              selections: transformedSysFields,
+            },
+          }
+
+          node.selections = node.selections
+            .map(field => {
+              if (SYS_FIELDS_TRANSFORMS.has(field.name?.value)) {
+                if (!sysInjected) {
+                  // inject for first occurence
+                  sysInjected = true
+                  return sysField
+                }
+                // remove all the others
+                return null
+              }
+              // keep non-sys fields as they are
+              return field
+            })
+            .filter(Boolean)
```

Comment:
@todo consider replacing this injection logic with the array splicing which was used for the asset fields